Refactor measure operator in a new pauli module

thierry-martinez opened the pull request:
This pull-request introduces a `pauli` module for symbolic Pauli calculations, which are used in `statevec` instead of direct matrix computation. The `test_pauli` module checks that the computations are the same as before, modulo a bug fix: ` CLIFFORD[CLIFFORD_CONJ[vop]]` was used instead of `CLIFFORD[vop].conj().T`, despite the two are only equal modulo phase in the general case (see the comment in `pauli.py`). The new implementation allows the angle and the plane to be computed separately from the state vector back-end, which will be useful for future client/server computations and to share the code with other back-ends.

shinich1 commented on graphix/pauli.py:
```diff
@@ -0,0 +1,285 @@
+    @property
+    def cos(self) -> Axis:
+        match self:
+            case Plane.XY:
+                return Axis.X
+            case Plane.YZ:
+                return Axis.Z  # former convention was Y
+            case Plane.XZ:
+                return Axis.Z  # former convention was X
+
+    @property
+    def sin(self) -> Axis:
+        match self:
+            case Plane.XY:
+                return Axis.Y
+            case Plane.YZ:
+                return Axis.Y  # former convention was Z
+            case Plane.XZ:
+                return Axis.X  # former convention was Z
+
+    def polar(self, angle: float) -> tuple[float, float, float]:
+        result = [0, 0, 0]
+        result[self.cos.value] = np.cos(angle)
+        result[self.sin.value] = np.sin(angle)
+        return tuple(result)
+
+    @staticmethod
+    def from_axes(a: Axis, b: Axis) -> "Plane":
+        if b.value < a.value:
+            a, b = b, a
+        match a, b:
+            case Axis.X, Axis.Y:
+                return Plane.XY
+            case Axis.Y, Axis.Z:
+                return Plane.YZ
+            case Axis.X, Axis.Z:
+                return Plane.XZ
+        assert a == b
+        raise ValueError(f"Cannot make a plane giving the same axis {a} twice.")
+
```
comment from @mgarnier59 : this convention is consistent with Backens et al., 
to @masa10-f : I think this is much clearer than what we had before- any comments?

masa10-f replied:
It seems nice. No comment so far :)

thierry-martinez commented:
Do we still want to support Python <3.10, or may we bump version constraint to allow `match` construct?

thierry-martinez commented:
> Are there significant differences? if we can work around with a few lines of additional code, my personal opinion is it's good to keep supporting 3.9. What's your thoughts?

Python 3.10 introduced `match`-blocks, allowing branching code to be written more elegantly by pattern-matching instead of `if-elif-else` chains. I rewrote in d7f7683 the few `match`-blocks I introduced into `if-elif-else` chains to be compatible with Python <3.10.

Python 3.9 allowed list and tuple types to be written as `list[...]` and `tuple[...]` instead of `typing.List[...]` and `typing.Tuple[...]`. I rewrote the types in 386f718 to be compatible with Python <3.9.

I think that when we will be ready to move to Python 3.10, `match`-blocks will allow many code places to be written in a more readable way, but there is no urgency to do so.

shinich1 commented:
> > Are there significant differences? if we can work around with a few lines of additional code, my personal opinion is it's good to keep supporting 3.9. What's your thoughts?
> 
> 
> 
> Python 3.10 introduced `match`-blocks, allowing branching code to be written more elegantly by pattern-matching instead of `if-elif-else` chains. I rewrote in d7f7683 the few `match`-blocks I introduced into `if-elif-else` chains to be compatible with Python <3.10.
> 
> 
> 
> Python 3.9 allowed list and tuple types to be written as `list[...]` and `tuple[...]` instead of `typing.List[...]` and `typing.Tuple[...]`. I rewrote the types in 386f718 to be compatible with Python <3.9.
> 
> 
> 
> I think that when we will be ready to move to Python 3.10, `match`-blocks will allow many code places to be written in a more readable way, but there is no urgency to do so.

Thanks, that seems pretty good, let me think a bit.  @masa10-f: let's discuss this next week.

shinich1 reviewed (approved):
Looks good to me! I will add docstring and docs page, once this PR is merged.
